### PR TITLE
Added a fix so that closed sockets do not cause errors

### PR DESCRIPTION
**Pull request by john-sharratt**

### CODE REVIEW

**Review by theduke — APPROVED** [review]

File: lib/virtual-net/src/tcp_pair.rs [hunk 1]
```diff
@@ -15,13 +15,21 @@ use std::{net::SocketAddr, task::Poll};
...
+    Shutdown,
+}
+
 #[derive(Debug)]
 struct SocketBufferState {
     buffer: RingBuffer<'static, u8>,
     push_handler: Option<ArcInterestHandler>,
     pull_handler: Option<ArcInterestHandler>,
     wakers: Vec<Waker>,
-    dead: bool,
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:24]: Fix this

File: lib/virtual-net/src/tcp_pair.rs [hunk 2]
```diff
@@ -15,13 +15,21 @@ use std::{net::SocketAddr, task::Poll};
...
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum State {
+    Alive,
+    Dead,
+    Closed,
+    Shutdown,
+}
+
 #[derive(Debug)]
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:26]: Fix this

File: lib/virtual-net/src/tcp_pair.rs [hunk 3]
```diff
@@ -42,8 +50,12 @@ pub(crate) struct SocketBuffer {
 
 impl Drop for SocketBuffer {
     fn drop(&mut self) {
-        if self.dead_on_drop {
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:45]: Fix this

Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:45]: Fix this

File: lib/virtual-net/src/tcp_pair.rs [hunk 4]
```diff
@@ -56,7 +68,7 @@ impl SocketBuffer {
                 push_handler: None,
                 pull_handler: None,
                 wakers: Vec::new(),
-                dead: false,
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:59]: Fix this

File: lib/virtual-net/src/tcp_pair.rs [hunk 5]
```diff
@@ -65,7 +77,7 @@ impl SocketBuffer {
 
     pub fn set_push_handler(&self, mut handler: ArcInterestHandler) {
         let mut state = self.state.lock().unwrap();
-        if state.dead {
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:68]: Fix this

File: lib/virtual-net/src/tcp_pair.rs [hunk 6]
```diff
@@ -76,7 +88,7 @@ impl SocketBuffer {
 
     pub fn set_pull_handler(&self, mut handler: ArcInterestHandler) {
         let mut state = self.state.lock().unwrap();
-        if state.dead {
```
Comment (BooBoo272) [inline lib/virtual-net/src/tcp_pair.rs:79]: Fix this

**Pull request merged**